akka-http-core/src/main/scala/akka/http/impl/engine/ws/Handshake.scala

```diff
@@ -57,69 +64,202 @@ private[http] object Handshake {
+        connection.exists(_.hasUpgrade) &&
+        version.exists(_.hasVersion(CurrentWebsocketVersion)) &&
+        key.exists(k ⇒ isValidKey(k.key))) {
+
+        val header = new UpgradeToWebsocketLowLevel {
+          def requestedProtocols: Seq[String] = supportedProtocols
+
+          def handle(handler: Either[Flow[FrameEvent, FrameEvent, Any], Flow[Message, Message, Any]], subprotocol: Option[String]): HttpResponse = {
+            require(subprotocol.forall(chosen ⇒ supportedProtocols.contains(chosen)),
+              s"Tried to choose invalid subprotocol '$subprotocol' which wasn't offered by the client: [${requestedProtocols.mkString(", ")}]")
+            buildResponse(key.get, handler, subprotocol)
+          }
+
+          def handleFrames(handlerFlow: Flow[FrameEvent, FrameEvent, Any], subprotocol: Option[String]): HttpResponse =
+            handle(Left(handlerFlow), subprotocol)
+
+          override def handleMessages(handlerFlow: Flow[Message, Message, Any], subprotocol: Option[String] = None): HttpResponse =
+            handle(Right(handlerFlow), subprotocol)
+        }
+        Some(header)
+      } else None
+    }
+
+    /*
+      From: http://tools.ietf.org/html/rfc6455#section-4.2.2
+
+      1.  A Status-Line with a 101 response code as per RFC 2616
+          [RFC2616].  Such a response could look like "HTTP/1.1 101
+          Switching Protocols".
+
+      2.  An |Upgrade| header field with value "websocket" as per RFC
+          2616 [RFC2616].
+
+      3.  A |Connection| header field with value "Upgrade".
+
+      4.  A |Sec-WebSocket-Accept| header field.  The value of this
+          header field is constructed by concatenating /key/, defined
+          above in step 4 in Section 4.2.2, with the string "258EAFA5-
+          E914-47DA-95CA-C5AB0DC85B11", taking the SHA-1 hash of this
+          concatenated value to obtain a 20-byte value and base64-
+          encoding (see Section 4 of [RFC4648]) this 20-byte hash.
+    */
+    def buildResponse(key: `Sec-WebSocket-Key`, handler: Either[Flow[FrameEvent, FrameEvent, Any], Flow[Message, Message, Any]], subprotocol: Option[String]): HttpResponse =
+      HttpResponse(
+        StatusCodes.SwitchingProtocols,
+        subprotocol.map(p ⇒ `Sec-WebSocket-Protocol`(Seq(p))).toList :::
+          List(
+            UpgradeHeader,
+            ConnectionUpgradeHeader,
+            `Sec-WebSocket-Accept`.forKey(key),
+            UpgradeToWebsocketResponseHeader(handler)))
+  }
+
+  object Client {
+    case class NegotiatedWebsocketSettings(subprotocol: Option[String])
+
+    /**
+     * Builds a WebSocket handshake request.
+     */
+    def buildRequest(uri: Uri, extraHeaders: immutable.Seq[HttpHeader], subprotocols: Seq[String], random: Random): (HttpRequest, `Sec-WebSocket-Key`) = {
+      val keyBytes = new Array[Byte](16)
+      random.nextBytes(keyBytes)
+      val key = `Sec-WebSocket-Key`(Base64.rfc2045().encodeToString(keyBytes, false))
```

Comment:
Would it make sense to put the `Base...` logic into its apply method (giving it a `keyBytes`)?